src/Module.zig

```diff
@@ -2113,6 +2109,15 @@ pub const File = struct {
 
     /// Add a reference to this file during AstGen.
     pub fn addReference(file: *File, mod: Module, ref: Reference) !void {
+        // Don't add the same module root twice
+        if (ref == .root) {
+            for (file.references.items) |other| {
+                if (other == .root and other.root == ref.root) {
+                    return;
+                }
+            }
+        }
```

Comment:
This O(N) lookup is sus. It seems to me that, instead of this logic, the burden should be on the calling code to not add the same module root twice.

Reply:
The call site in question is `importPkg` - a new call made it possible to error for a file being the root of two modules (which was accidentally left out before but that was a good thing because it would have broken loads of dep patterns that accidentally worked with `--pkg-begin`). There's not really any way it can know whether to add the reference again without just doing this traversal.
Now I think about it, one thing we could do is always store "root-of" references at the front of this list. Those references are reasonably bounded in practice, and we reorder the list to that order when we emit the errors anyway anyway, We only need to check against those, so we can put them at the front and break when we see a `.import`. That'd make this loop O(1) on valid code (since there'll only be one "root-of" reference) and not realistically much slower on bad code. I'll go for that approach unless you have any better ideas.

Reply:
O(1) on valid code, now that's what I like to see!

Reply:
To be fair now I think about it this was O(1) on valid code anyway, since you either have at most one reference (if it's a module root) or all the references are "imported-from-file" - still, the new one is nicer